runtime/doc/lua.txt

```diff
@@ -2773,4 +2870,367 @@ range({spec})                                            *vim.version.range()*
+    Add a filter step to the iterator pipeline.
+
+    Example: >lua
+
+     local bufs = vim.iter(vim.api.nvim_list_bufs()):filter(vim.api.nvim_buf_is_loaded)
+<
+
+    Parameters: ~
+      • {f}  function(...):bool Takes all values returned from the previous
+             stage in the pipeline and returns false or nil if the current
+             iterator element should be removed.
+
+    Return: ~
+        Iter
+
+Iter:filtermap({self}, {f})                                 *Iter:filtermap()*
+    Add a map and filter step to the iterator pipeline.
+
+    Example: >lua
+
+     local it = vim.iter({ 1, 2, 3, 4 }):filtermap(function(v)
+       if v % 2 == 0 then
+         return v * 3
+       end
+     end)
+     it:collect()
+     -- { 6, 12 }
+<
+
+    Parameters: ~
+      • {f}  function(...):any Mapping function. Takes all values returned
+             from the previous stage in the pipeline as arguments and returns
+             one or more new values, which are used in the next pipeline
+             stage. Nil return values returned are filtered from the output.
+
+    Return: ~
+        Iter
+
+Iter:find({self}, {f})                                           *Iter:find()*
+    Find the first value in the iterator that satisfies the given predicate.
+
+    Advances the iterator. Returns nil and drains the iterator if no value is
+    found.
+
+    Examples: >lua
+
+     local it = vim.iter({ 3, 6, 9, 12 })
+     it:find(12)
+     -- 12
+
+     local it = vim.iter({ 3, 6, 9, 12 })
+     it:find(20)
+     -- nil
+
+     local it = vim.iter({ 3, 6, 9, 12 })
+     it:find(function(v) return v % 4 == 0 end)
+     -- 12
+<
+
+    Return: ~
+        any
+
+Iter:foreach({self}, {f})                                     *Iter:foreach()*
+    Call a function once for each item in the pipeline.
+
+    This is used for functions which have side effects. To modify the values
+    in the iterator, use |Iter:filtermap()|.
+
+    This function drains the iterator.
+
+    Parameters: ~
+      • {f}  function(...) Function to execute for each item in the pipeline.
+             Takes all of the values returned by the previous stage in the
+             pipeline as arguments.
+
+Iter:last({self})                                                *Iter:last()*
+    Return the last item in the iterator.
+
+    Drains the iterator.
+
+    Example: >lua
+
+     local it = vim.iter(vim.gsplit('abcdefg', ''))
+     it:last()
+     -- 'g'
+
+     local it = vim.iter({ 3, 6, 9, 12, 15 })
+     it:last()
+     -- 15
+<
+
+    Return: ~
+        any
+
+Iter:next({self})                                                *Iter:next()*
+    Return the next value from the iterator.
+
+    Example: >lua
+
+     local it = vim.iter(string.gmatch('1 2 3', 'd+')):filtermap(tonumber)
+     it:next()
+     -- 1
+     it:next()
+     -- 2
+     it:next()
+     -- 3
+<
+
+    Return: ~
+        any
+
+Iter:nextback({self})                                        *Iter:nextback()*
```

Comment:
On a related note, `revnext` ("reverse next") or `endnext` might be better -- unless `nextback` is already an established term in iteratorland -- as "back" isn't used anywhere when talking about iterators here.

Alternatively, isn't that just `:rev():next()`? Do we need a separate function for this?

Reply:
(First point applies to all the `fooback` functions  -- the docs always talk about the "end of the iterator" or the "last element".)

Reply:
> On a related note, `revnext` ("reverse next") or `endnext` might be better -- unless `nextback` is already an established term in iteratorland -- as "back" isn't used anywhere when talking about iterators here.

"back" is a standard term in at least [Rust](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html#tymethod.next_back) and [C++](https://en.cppreference.com/w/cpp/container/vector/back).

> Alternatively, isn't that just `:rev():next()`? Do we need a separate function for this?

No, it's not quite the same. `:rev()` will reverse the order of the table. One may want to remove a value from the end of the list without reversing the order of the list.